projects/packages/jetpack-mu-wpcom/src/features/launchpad/launchpad-task-definitions.php

```diff
@@ -394,35 +394,88 @@ function wpcom_launchpad_track_completed_task( $task_id, $extra_props = array()
+		}
+		$mapping[ $value['id_map'] ] = $task_id;
 	}
 
-	// If the task has an id_map, use that instead.
-	$key = $task_id;
-	if ( isset( $task_definitions[ $task_id ]['id_map'] ) ) {
-		$key = $task_definitions[ $task_id ]['id_map'];
-	}
+	return $mapping;
+}
+/**
+ * Mark a task as complete.
+ *
+ * @param string|string[] $task_ids The task IDs that should be marked as complete.
+ * @return bool True if the task was marked as complete, false otherwise.
+ */
+function wpcom_mark_launchpad_task_complete( $task_ids ) {
+	$task_ids = is_array( $task_ids ) ? $task_ids : array( $task_ids );
+	$values   = array_fill_keys( $task_ids, true );
 
-	$statuses         = get_option( 'launchpad_checklist_tasks_statuses', array() );
-	$statuses[ $key ] = true;
-	$result           = update_option( 'launchpad_checklist_tasks_statuses', $statuses );
+	$result = wpcom_update_launchpad_task_status( $values );
 
 	// Record the completion event in Tracks.
-	wpcom_launchpad_track_completed_task( $key );
+	foreach ( $task_ids as $key => $value ) {
+		wpcom_launchpad_track_completed_task( $key );
+	}
 
 	return $result;
 }
 
+/**
+ * Mark a task as incomplete. This method is here for backwards compatibility of
+ * the Rest API that allowed marking a task as incomplete.
+ *
+ * @param string|string[] $task_ids The task IDs that should be marked as complete.
+ * @return bool True if the task was marked as complete, false otherwise.
+ */
+function wpcom_mark_launchpad_task_incomplete( $task_ids ) {
+	$task_ids = is_array( $task_ids ) ? $task_ids : array( $task_ids );
+	$values   = array_fill_keys( $task_ids, false );
+
+	return wpcom_update_launchpad_task_status( $values );
+}
+
+/**
+ * Updates task/s statuses. If a non-existent task is passed, it will be ignored.
+ *
+ * @param array $original_statuses Array of mappings [ task_id: string => status: bool ].
+ * @return bool True if the option where the statuses are stored was updated, false otherwise.
+ */
+function wpcom_update_launchpad_task_status( $original_statuses ) {
+	$task_definitions = wpcom_launchpad_get_task_definitions();
+	$reverse_id_map   = wpcom_launchpad_get_reverse_id_mappings();
+
+	$statuses = array();
+	foreach ( $original_statuses as $task_id => $value ) {
+		// Find out wether the task id sent is the actual task id or the id_map.
+		$actual_task_id = ( isset( $reverse_id_map[ $task_id ] ) ) ? $reverse_id_map[ $task_id ] : $task_id;
+		if ( ! isset( $task_definitions[ $actual_task_id ] ) ) {
+			// Bail if no task definition was found.
+			continue;
+		}
+
+		// If the task has an id_map, use that, otherwise use the actual task id as the key.
+		$task                        = $task_definitions[ $actual_task_id ];
+		$stored_task_id              = isset( $task['id_map'] ) ? $task['id_map'] : $actual_task_id;
+		$statuses[ $stored_task_id ] = $value;
+	}
+
+	$old_values = (array) get_option( 'launchpad_checklist_tasks_statuses', array() );
+	$new_values = array_merge( $old_values, $statuses );
```

Comment:
As per my previous comments, it may be worth adding the tracking logic in this block of code. I think we also want to remove the values when the new value is false - should we filter out non-true values?
```suggestion
	$old_status_values = (array) get_option( 'launchpad_checklist_tasks_statuses', array() );
	$statuses = array();
	foreach ( $new_statuses as $task_id => $value ) {
		// Find out wether the task id sent is the actual task id or the id_map.
		$actual_task_id = ( isset( $reverse_id_map[ $task_id ] ) ) ? $reverse_id_map[ $task_id ] : $task_id;
		if ( ! isset( $task_definitions[ $actual_task_id ] ) ) {
			// Bail if no task definition was found.
			continue;
		}

		// If the task has an id_map, use that, otherwise use the actual task id as the key.
		$task                        = $task_definitions[ $actual_task_id ];
		$stored_task_id              = isset( $task['id_map'] ) ? $task['id_map'] : $actual_task_id;
		$statuses[ $stored_task_id ] =(bool) $value;

		// If we're completing a task for the first time, track that completion via the supplied task ID.
		if ( $value && empty( $old_status_values[ $stored_task_id ] ) ) {
			wpcom_launchpad_track_completed_task( $task_id );
		}
	}

	$new_values = array_merge( $old_values, $statuses );
	// Consider adding a filter for non-truthy values
	$new_values = array_filter( $new_values );
```

Reply:
I've fixed this but I didn't add the `array_filter` in the end to prevent having to unset an existing key if the task is being marked as incomplete.